lib/Dialect/FIRRTL/Transforms/InferWidths.cpp

```diff
@@ -0,0 +1,770 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the InferWidths pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "./PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "llvm/ADT/DenseSet.h"
+#include "llvm/ADT/GraphTraits.h"
+#include "llvm/ADT/Hashing.h"
+#include "llvm/ADT/MapVector.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "infer-widths"
+
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Constraint Expressions
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct Expr;
+} // namespace
+
+/// Allow rvalue refs to `Expr` and subclasses to be printed to streams.
+template <typename T, typename std::enable_if<std::is_base_of<Expr, T>::value,
+                                              int>::type = 0>
+inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os, const T &e) {
+  e.print(os);
+  return os;
+}
+
+/// Allow pointers to `Expr` and subclasses to be printed to streams.
+template <typename T, typename std::enable_if<std::is_base_of<Expr, T>::value,
+                                              int>::type = 0>
+inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os, const T *e) {
+  return os << *e;
+}
+
+// Allow expression subclasses to be hashed.
+namespace mlir {
+template <typename T, typename std::enable_if<std::is_base_of<Expr, T>::value,
+                                              int>::type = 0>
+inline llvm::hash_code hash_value(const T &x) {
+  return x.hash_value();
+}
+} // namespace mlir
+
+namespace {
+
+/// An expression on the right-hand side of a constraint.
+struct Expr {
+  enum Kind {
+    KindRoot,
+    KindVar,
+    KindKnown,
+    KindAdd,
+    KindMax,
+    KindMin,
+  };
+  Kind kind;
+  llvm::Optional<int32_t> solution = {};
+
+  /// Print a human-readable representation of this expr.
+  void print(llvm::raw_ostream &os) const;
+
+  // Iterators over the child expressions.
+  typedef Expr *const *iterator;
+  iterator begin() const;
+  iterator end() const;
+
+protected:
+  Expr(Kind kind) : kind(kind) {}
+  llvm::hash_code hash_value() const { return llvm::hash_value(kind); }
+};
+
+/// Helper class to CRTP-derive common functions.
+template <class DerivedT, Expr::Kind DerivedKind>
+struct ExprBase : public Expr {
+  ExprBase() : Expr(DerivedKind) {}
+  static bool classof(const Expr *e) { return e->kind == DerivedKind; }
+  bool operator==(const Expr &other) const {
+    if (auto otherSame = dyn_cast<DerivedT>(other))
+      return *this == otherSame;
```

Comment:
Does this reliably work without recursion?

Reply:
I think a `static_cast<DerivedT *>` wouldn't hurt here, to ensure we don't accidentally end up in an infinite recursion. Added that.